src/repository.rs

```diff
@@ -82,57 +175,334 @@ where
     }
 }
 
-impl<T, D> Repository<D> for &T
+impl<R, D> Repository<R, D>
 where
-    T: Repository<D>,
+    R: RepositoryProvider,
     D: DataInterchange + Sync,
 {
-    fn store_metadata<'a, M>(
+    /// Fetch and parse metadata identified by `meta_path`, `version`, and
+    /// [`D::extension()`][extension].
+    ///
+    /// If `max_length` is provided, this method will return an error if the metadata exceeds
+    /// `max_length` bytes. If `hash_data` is provided, this method will return and error if the
+    /// hashed bytes of the metadata do not match `hash_data`.
+    ///
+    /// [extension]: crate::interchange::DataInterchange::extension
+    pub async fn fetch_metadata<'a, M>(
         &'a self,
         meta_path: &'a MetadataPath,
         version: &'a MetadataVersion,
-        metadata: &'a SignedMetadata<D, M>,
-    ) -> BoxFuture<'a, Result<()>>
+        max_length: Option<usize>,
+        hash_data: Option<(&'static HashAlgorithm, HashValue)>,
+    ) -> Result<(RawSignedMetadata<D, M>, SignedMetadata<D, M>)>
     where
-        M: Metadata + Sync + 'static,
+        M: Metadata,
     {
-        (**self).store_metadata(meta_path, version, metadata)
+        let raw_signed_meta = self
+            .fetch_raw_metadata(meta_path, version, max_length, hash_data)
+            .await?;
+        let signed_meta = raw_signed_meta.parse()?;
```

Comment:
The match statements in lookup_target_description get rather nasty if there isn't a single method I can call to get both.

Reply:
I've got a little concern that if we ever have two `fetch_metadata` calls in the same scope, it'd be a little easier to hash check one metadata, but use the other when trying to update the tuf database. This probably isn't a large problem in practice since we still go through signature verification and the rest of the validation, but it'd be nice to avoid this.

We can defer addressing this for a future patch though.